src/debug/telemetry/telemetry2_slot.c

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (c) 2024 Intel Corporation
+
+#include <zephyr/logging/log.h>
+#include <zephyr/spinlock.h>
+#include <adsp_debug_window.h>
+#include <adsp_memory.h>
+#include <sof/common.h>
+#include "telemetry2_slot.h"
+
+LOG_MODULE_REGISTER(telemetry2_slot);
+
+static struct k_spinlock lock;
+
+/*
+ * Search, or if not found reserve, a chunk of specified size and id from
+ * telemetry2 slot. Each chuck is forced to star from 64-byte cache line
+ * boundary. A pointer to the chunk is returned.
+ */
+void *telemetry2_chunk_get(int id, size_t required_size)
+{
+	int slot = CONFIG_SOF_TELEMETRY2_SLOT_NUMBER;
+	struct telemetry2_payload_hdr *payload =
+		(struct telemetry2_payload_hdr *)(ADSP_DW->slots[slot]);
+	size_t hdr_size = ALIGN_UP(sizeof(*payload), 64);
+	struct telemetry2_chunk_hdr *chunk = (struct telemetry2_chunk_hdr *)
+		(((uint8_t *) payload) + hdr_size);
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	size_t size = ALIGN_UP(required_size, 64);
+	uint8_t *p;
+
+	if (ADSP_DW->descs[slot].type != ADSP_DW_SLOT_TELEMETRY2) {
+		if (ADSP_DW->descs[slot].type != 0)
+			LOG_WRN("Slot %d was not free: %u", slot, ADSP_DW->descs[slot].type);
+		LOG_INF("Initializing telemetry2 slot 0x%08x", ADSP_DW->descs[slot].type);
+		ADSP_DW->descs[slot].type = ADSP_DW_SLOT_TELEMETRY2;
+		payload->hdr_size = hdr_size;
+		payload->magic = TELEMETRY2_PAYLOAD_MAGIC;
+		payload->abi = TELEMETRY2_PAYLOAD_V0_0;
+	}
+
+	LOG_INF("Add id %u size %u (after alignment %u)", id, required_size, size);
+	while (chunk->id != id && chunk->size != 0) {
+		p = (((uint8_t *) chunk) + chunk->size);
+		if (p + size >= ADSP_DW->slots[slot + 1]) {
+			LOG_ERR("No space for chunk %u of size %u in slot %u, offset %u",
+				id, size, slot, p - ADSP_DW->slots[slot]);
+			goto err;
+		}
+
+		chunk = (struct telemetry2_chunk_hdr *) p;
+	}
+
+	if (id == chunk->id) {
+		if (required_size > chunk->size) {
+			LOG_ERR("Chunk %d size too small %u > %u",
+				id, required_size, chunk->size);
+			goto err;
+		}
+		LOG_INF("Chunk %u found", id);
+		goto out;
+	}
+
+	if (chunk->id != TELEMETRY2_CHUNK_ID_EMPTY)
+		LOG_WRN("Chunk of size %u has type %u, assuming empty",
+			chunk->size, chunk->id);
+
+	LOG_INF("Chunk %u reserved", id);
+	chunk->size = size;
+	chunk->id = id;
+	payload->total_size = size + ((uint8_t *) chunk) - ((uint8_t *) payload);
+
+out:
+	k_spin_unlock(&lock, key);
```

Comment:
do we need to hold this lock (with IRQs off) for this long ?

Reply:
I think I could make a version that first checks without locks if everything is already initialized and reserved, and only take the lock if that is not the case. This way the lock would be taken only once per reserved chunk. I did not go through the trouble since the the code is anyway run only  once per thread accessing the telemetry2 chunks.